meinberlin/apps/votes/forms.py

```diff
@@ -1,34 +1,62 @@
 from django import forms
+from django.forms.widgets import TextInput
 from django.utils.translation import gettext_lazy as _
 
 from meinberlin.apps.votes.models import VotingToken
 
 
-class VotingTokenWidget(forms.MultiWidget):
+class NumberInput(TextInput):
 
-    def __init__(self, attrs=None):
-        widgets = (
-            forms.TextInput(attrs={'minlength': '4',
-                                   'maxlength': '4'}),
-            forms.TextInput(attrs={'minlength': '4',
-                                   'maxlength': '4'}),
-            forms.TextInput(attrs={'minlength': '4',
-                                   'maxlength': '4'}),
-        )
-        super(VotingTokenWidget, self).__init__(widgets, attrs)
+    # switch input type to type tel so numeric keyboard shows on mobile devices
+    input_type = 'tel'
 
-    def decompress(self, value):
-        if value:
-            return [value[i:i + 4] for i in range(0, len(value), 4)]
-        return [None, None, None]
 
-    def value_from_datadict(self, data, files, name):
-        values = super().value_from_datadict(data, files, name)
-        return "".join(values)
+class VotingTokenWidget(forms.CharField):
+    token = {
+        'pattern': [5152, 5354, 5522],
+        'length': [12]
+    }
+
+    def __init__(self, placeholder=None, *args, **kwargs):
+        widget = (NumberInput(attrs={'minlength': '12',
+                                     'maxlength': '12',
+                                     'placeholder': placeholder})
+                  )
+        super(VotingTokenWidget, self).__init__(widget=widget, *args, **kwargs)
+
+    default_error_messages = {
+        'invalid': _('The token is invalid'),
+    }
+
+    def clean(self, value):
+
+        # ensure no spaces or dashes
+        value = value.replace(' ', '').replace('-', '')
+
+        # get the token type and its specs
+        token = self.token_from_number(value)
+
+        # if no token found, invalid
+        if not token:
+            raise forms.ValidationError(self.error_messages['invalid'])
+
+        # check the length
+        if not len(value) in token['length']:
+            raise forms.ValidationError(self.error_messages['invalid'])
+
+        return value
+
+    def token_from_number(self, num):
+        # find this token, based on the token number
+        for token in self.token:
+            for pattern in token['patterns']:
+                if (str(pattern) == str(num)[:len(str(pattern))]):
+                    return token
```

Comment:
@fuzzylogic2000 here it is if you want to take a look, I will keep poking till lunch then maybe go on to react update test branch

Reply:
Right, I can also fix the (backend) test! Maybe when we are done with everything here?